tachyon/zk/lookup/lookup_verification.h

```diff
@@ -0,0 +1,78 @@
+// Copyright 2020-2022 The Electric Coin Company
+// Copyright 2022 The Halo2 developers
+// Use of this source code is governed by a MIT/Apache-2.0 style license that
+// can be found in the LICENSE-MIT.halo2 and the LICENCE-APACHE.halo2
+// file.
+
+#ifndef TACHYON_ZK_LOOKUP_LOOKUP_VERIFICATION_H_
+#define TACHYON_ZK_LOOKUP_LOOKUP_VERIFICATION_H_
+
+#include <memory>
+#include <vector>
+
+#include "tachyon/zk/lookup/lookup_argument.h"
+#include "tachyon/zk/lookup/lookup_verification_data.h"
+#include "tachyon/zk/plonk/vanishing/vanishing_verification_evaluator.h"
+
+namespace tachyon::zk {
+
+template <typename F>
+F CompressExpressions(
+    const std::vector<std::unique_ptr<Expression<F>>>& expressions,
+    const F& theta, VanishingVerificationEvaluator<F>& evaluator) {
+  F compressed_value = F::Zero();
+  for (size_t expr_idx = 0; expr_idx < expressions.size(); ++expr_idx) {
+    compressed_value *= theta;
+    compressed_value += evaluator.Evaluate(expressions[expr_idx].get());
+  }
+  return compressed_value;
+}
+
+template <typename F>
+F CreateProductExpression(const LookupVerificationData<F>& data,
+                          const LookupArgument<F>& argument) {
+  VanishingVerificationEvaluator<F> evaluator(data);
+  // z(ω * X) * (a'(X) + β) * (s'(X) + γ)
+  // - z(X) * (θᵐ⁻¹a₀(X) + ... + aₘ₋₁(X) + β) * (θᵐ⁻¹s₀(X) + ... + sₘ₋₁(X) + γ)
+  F left = data.product_next_eval * (data.permuted_input_eval + data.beta) *
+           (data.permuted_table_eval + data.gamma);
+  F compressed_input_expression =
+      CompressExpressions(argument.input_expressions(), data.theta, evaluator);
+  F compressed_table_expression =
+      CompressExpressions(argument.table_expressions(), data.theta, evaluator);
+  F right = data.product_eval * (compressed_input_expression + data.beta) *
+            (compressed_table_expression + data.gamma);
+  return left - right;
```

Comment:
```suggestion
  return left -= right;
```

Reply:
Since this function returns `F`, so even though you're trying to `-=` it occurs copy in the end.